src/pkg/packager/deploy.go

```diff
@@ -118,39 +148,86 @@ func (p *Packager) deployComponents() (deployedComponents []types.DeployedCompon
+			deployedComponents[idx].InstalledCharts = installedCharts
+		}
+
+		// Update the package secret to indicate that we are attempting to deploy this component
+		if p.cluster != nil && !p.cfg.IsInitConfig {
+			if _, err = p.cluster.RecordPackageDeploymentAndWait(p.cfg.Pkg, deployedComponents, p.connectStrings, p.generation); err != nil {
+				message.Warnf("Unable to record package deployment for component %s: this will affect features like `zarf package remove`: %s", component.Name, err.Error())
+			}
+		}
+
+		// Deploy the component
+		var charts []types.InstalledChart
+		var deployErr error
 		if p.cfg.IsInitConfig {
-			charts, err = p.deployInitComponent(component)
+			charts, deployErr = p.deployInitComponent(component)
 		} else {
-			charts, err = p.deployComponent(component, false /* keep img checksum */, false /* always push images */)
+			charts, deployErr = p.deployComponent(component, false /* keep img checksum */, false /* always push images */)
 		}
 
-		deployedComponent := types.DeployedComponent{Name: component.Name}
 		onDeploy := component.Actions.OnDeploy
 
 		onFailure := func() {
 			if err := p.runActions(onDeploy.Defaults, onDeploy.OnFailure, p.valueTemplate); err != nil {
 				message.Debugf("unable to run component failure action: %s", err.Error())
 			}
 		}
-		if err != nil {
+
+		if deployErr != nil {
 			onFailure()
-			return deployedComponents, fmt.Errorf("unable to deploy component %s: %w", component.Name, err)
+
+			deployedComponents[idx].Status = types.ComponentStatusFailed
+
+			if p.cluster != nil {
+				_, err = p.cluster.RecordPackageDeploymentAndWait(p.cfg.Pkg, deployedComponents, p.connectStrings, p.generation)
+				if err != nil && deployedComponent.Name != "zarf-injector" {
+					message.Warnf("Unable to record package deployment for component %s: this will affect features like `zarf package remove`: %s", component.Name, err.Error())
+				}
+			}
+
+			return deployedComponents, fmt.Errorf("unable to deploy component %s: %w", component.Name, deployErr)
 		}
 
-		// Deploy the component
-		deployedComponent.InstalledCharts = charts
-		deployedComponents = append(deployedComponents, deployedComponent)
+		// Update the deployed component secret
+		deployedComponents[idx].InstalledCharts = charts
+		deployedComponents[idx].Status = types.ComponentStatusSucceeded
 
 		// Save deployed package information to k8s
 		// Note: Not all packages need k8s; check if k8s is being used before saving the secret
 		if p.cluster != nil {
-			err = p.cluster.RecordPackageDeployment(p.cfg.Pkg, deployedComponents, p.connectStrings)
-			if err != nil {
-				message.Debugf("Unable to record package deployment for component %s: this will affect features like `zarf package remove`: %s", component.Name, err.Error())
+			_, err = p.cluster.RecordPackageDeploymentAndWait(p.cfg.Pkg, deployedComponents, p.connectStrings, p.generation)
+			if err != nil && deployedComponent.Name != "zarf-injector" {
```

Comment:
We should not hard code init component names (working to remove the ones that do exist)